src/models/model.ts

```diff
@@ -0,0 +1,44 @@
+import { Attribute } from './attributes';
+
+export type SaveCallback = (error: Error | null, result?: any) => void;
+
+export default class Model {
+  static attributes: { [key: string]: Attribute };
+
+  constructor(json?: any) {
+    if (json) {
+      this.fromJSON(json);
+    }
+  }
+
+  attributes(): { [key: string]: Attribute } {
+    return (this.constructor as any).attributes;
+  }
+
+  fromJSON(json?: any): Model {
+    const attributes = this.attributes();
+    for (const attrName in attributes) {
+      const attr = attributes[attrName];
+      if (json[attr.jsonKey] !== undefined) {
+        (this as any)[attrName] = attr.fromJSON(json[attr.jsonKey], this);
```

Comment:
I think you should be able to have `(this as Model)` here

Reply:
I tried that but it throws an error complaining about using a `string` as a index for type `Model`

> TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'Model'. No index signature with a parameter of type 'string' was found on type 'Model'.

One way around this would be to have an interface like `export interface ModelProperties { [key: string]: any }` and use it like: `(this as ModelProperties)[attrName] = attr.fromJSON(json[attr.jsonKey], this);` but I'm not sure if that's any better.